### DISCUSSION TITLE
Dynamically Adding Query Mods To ThenLoad

### DISCUSSION THREAD

**Discussion by daddz** [stephenafamo/bob#173] [Ideas]
A lot of times I need to conditionally add one or more query mods to `ThenLoad*`-ers - those might also be nested a few levels but that's not as relevant here.

Currently the Loaders cannot be changed after creation so I did wrap them in simplistic way so I can build them up as needed which looks something like this:

```go
type LoaderFunc func(queryMods ...bob.Mod[*dialect.SelectQuery]) psql.Loader

type ThenLoad struct {
	loader LoaderFunc
	mods   []bob.Mod[*dialect.SelectQuery]
}

func NewThenLoad(loaderFunc LoaderFunc) *RelationLoader {
	return &RelationLoader{
		loader: loaderFunc,
	}
}

func (s *ThenLoad) Append(clause bob.Mod[*dialect.SelectQuery]) {
	s.mods = append(s.mods, clause)
}

func (s *ThenLoad) Loader() psql.Loader {
	return s.loader(s.mods...)
}
```

Then it can be used like this:
```go

relation := NewThenLoad(models.ThenLoadFooBar)
relation.Append(models.ThenLoadBarBaz)

if someCondition {
  relation.Append(models.SelectWhere.Bar.Field.EQ(someVal))
}

// example without any other query mods besides the relation loader
result, _ := models.Foo.Query(ctx, db, relation.Loader())
```

I think this use case is not that unusual.

Would it make sense to add such a capability to the `ThenLoad` directly?

**Comment by stephenafamo** [comment]
Maybe I'm missing something, but can't you achieve the same by keeping a slice of mods and then applying it?

```go
var relationMods []bob.Mod[*dialect.SelectQuery]

if someCondition {
  relationMods = append(relationMods, models.SelectWhere.Bar.Field.EQ(someVal))
}

// example without any other query mods besides the relation loader
result, _ := models.Foo.Query(ctx, db, models.ThenLoadFooBar(relationMods...))
```

**Reply by daddz** [reply]
Yes you're right, I might have stripped down the example to be too simplistic :)

It can get rather unwieldy when dealing with multiple relations that are also nested and you need different variations for e.g. a `Count()` and the actual `All()` query. It's more like a developer experience improvement suggestion.

I'll play around with different ideas a bit more and might come back with a more fleshed out suggestion.

Thanks for your quick feedback!
